lineage/states/StateDistribution.py

```diff
@@ -87,25 +97,28 @@ def __repr__(self):
 # can handle the case where the list of observations is empty.
 
 
-def gamma_estimator(gamma_obs):
+def gamma_estimator(gamma_obs, gamma_censor_obs, old_params, gammas):
     """
     This is a closed-form estimator for two parameters
     of the Gamma distribution, which is corrected for bias.
     """
-    N = len(gamma_obs)
-
-    xbar = (sum(gamma_obs) + 49e-10) / (len(gamma_obs) + 1e-10)
-    x_lnx = [x * np.log(x) for x in gamma_obs]
-    lnx = [np.log(x) for x in gamma_obs]
-    # gamma_a
-    a_hat = (N * (sum(gamma_obs)) + 7e-10) / (N * sum(x_lnx) - (sum(lnx)) * (sum(gamma_obs)) + 1e-10)
-    # gamma_scale
-    b_hat = xbar / a_hat
+    s = np.log(sum(gammas*gamma_obs)/sum(gammas)) - sum(gammas*np.log(gamma_obs))/sum(gammas)
+    k0 = old_params[0]
+    scale0 = old_params[1]
+    
+    f = lambda k,s : np.log(k) - sc.polygamma(0,k) - s
+    fprime = lambda k,s : (1./k) - sc.polygamma(1,k)
+    fprime2 = lambda k,s : (-1./k**2) - sc.polygamma(2,k)
+    
+    halley = lambda k,s: k - ((2*(f(k,s)*fprime(k,s)))/(2*(fprime(k,s)**2)-f(k,s)*fprime2(k,s)))
+    a_hat = halley(k0, s)
 
-    if b_hat < 1.0 or 50.0 < a_hat < 5.0:
-        return 7, 7
+    scale_hat = (sum(gammas*gamma_obs))/a_hat/sum(gammas)
+    
+    if a_hat > 100:
+        return 7, 3+(1*(np.random.uniform()))
 
-    return a_hat, b_hat
+    return a_hat, scale_hat
```

Comment:
example of how the `gammas` or the participation variable, or the component or state identifier variable is now utilized in the estimation of the parameters themselves.

halley root finder is now used.

Reply:
Aren't you only doing one iteration of Halley here? It's iterative and so would need to run until convergence.

If you pass the `fprime2` argument here, scipy can take care of Halley's method and convergence checks:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html#scipy.optimize.newton

Reply:
I tried using the Newton root finder originally (it was the documentation for that that led me to Halley's method) but the issue I had with it was it would create a Runtime error when it wasn't converging. When I tried it in a notebook I was getting the Halley's method to converge in 1 to 2 iterations, so I thought this would be better.